Add typing extensions

Michael0x2a opened the pull request:
This pull request adds a `typing_extensions` subproject to `typing`. The `typing_extensions` module backports any new additions to `typing` for Python 3.5+ users who are using older versions of `typing` that were bundled with their standard library (and so can't update to the latest versions).

See https://github.com/python/typing/issues/435 for more motivation and additional context.

ilevkivskyi commented:
I don't think you need all these old versions bundled in tests. I think we should just properly configure Travis (it supports Python micro releases like 3.5.2 etc. (like in mypy's ``.travis.yml``), and specific versions of ``typing`` to be installed before tests)

Michael0x2a commented:
@ilevkivskyi I guess my only concern with that is that it becomes harder to do sanity checks locally -- having to constantly push to run tests seems a bit annoying to me.

ilevkivskyi commented:
> having to constantly push to run tests seems a bit annoying to me.

You can still keep these copies locally (without tracking them in git). But I would prefer a properly configured ``.travis.yml`` to test with actual Python versions (note that Python runtime changes itself between versions, not only the stdlib, there are few places in ``typing`` where this is actually important).

ilevkivskyi commented on typing_extensions/src_py3/typing_extensions.py:
```diff
@@ -0,0 +1,612 @@
+T_contra = typing.TypeVar('T_contra', contravariant=True)  # Ditto contravariant.
+
+
+def _gorg(a):
```
Note that ``_gorg`` and ``_geqv`` will be removed soon after Python 3.6.2 is released, see PR https://github.com/python/typing/pull/439.

Michael0x2a replied:
Sounds good -- once this pull request gets accepted, I can prepare another pull request to add those changes and leave it pending just like #439.

ilevkivskyi replied:
> or whatever other course of action you think is best

OK, let's just not forget about this.

ilevkivskyi commented:
Concerning the test failure, we decided to stop supporting Python 3.2 (it reached EoL a year ago), so that you can just skip testing on Python 3.2.

Michael0x2a commented:
Update: I removed `run_tests.py` and the associated machinery in favor of just using Travis, as suggested. The only remaining blocker is that we can't actually install typing on Python 3.2. I figured it'd be best to make a separate pull request for that though: https://github.com/python/typing/pull/444

Once that's accepted, I can rebase, but this pull request should otherwise be ready for review (?).

ilevkivskyi commented:
I can review this, but I am not sure I will be able to review before week-end.

ilevkivskyi commented on typing_extensions/src_py2/typing_extensions.py:
```diff
@@ -0,0 +1,113 @@
+    'NewType',
+    'overload',
+    'Text',
+    'TYPE_CHECKING',
```
``GenericMeta`` was made public recently (in 3.6.1 IIRC), should we also add it here?

Michael0x2a replied:
I'm not sure if we should -- GenericMeta seems to have gone through some heavy changes since 3.5.0, and if we make it public here, we're basically guaranteeing it'll work reasonably consistently across all Python versions, and I'm not sure if it'll be easy for us to confirm that's really the case.

If you think it's still worth including though, I can add some tests for it + add it to `__all__`.

ilevkivskyi replied:
OK, let us not include ``GenericMeta`` for now. People who want to use it can just import it from ``typing`` (it just does not work with ``import *``).

ilevkivskyi commented on typing_extensions/src_py3/test_typing_extensions.py:
```diff
@@ -0,0 +1,636 @@
+
+TYPING_V2 = sys.version_info >= (3, 5, 1)
+TYPING_V3 = sys.version_info >= (3, 5, 3)
+TYPING_V4 = sys.version_info >= (3, 6, 1)
```
We will have ``TYPING_V5`` very soon (but I think it is OK to not include it in this PR, just don't forget to add it later).

Michael0x2a replied:
Will do.

ilevkivskyi commented on typing_extensions/src_py3/test_typing_extensions.py:
```diff
@@ -0,0 +1,636 @@
+try:
+    import collections.abc as collections_abc
+except ImportError:
+    import collections as collections_abc  # Fallback for PY3.2.
```
We don't support 3.2 anymore :-) this can be removed.

Michael0x2a replied:
Done!

ilevkivskyi commented on typing_extensions/src_py3/test_typing_extensions.py:
```diff
@@ -0,0 +1,636 @@
+                message += ' : %s' % msg
+            raise self.failureException(message)
+
+    def clear_caches(self):
```
Again, I am not sure you need this.

Michael0x2a replied:
Yep, removed.

ilevkivskyi commented on typing_extensions/src_py3/test_typing_extensions.py:
```diff
@@ -0,0 +1,636 @@
+        with self.assertRaises(TypeError):
+            issubclass(Employee, NoReturn)
+
+    @skipUnless(TYPING_V3, "Behavior added in typing v3")
```
It is not necessary, but it would be super-helpful to mention PR numbers where behavior was changed (also for other changes).

Michael0x2a replied:
I'll look into digging those up.

ilevkivskyi commented on typing_extensions/src_py3/test_typing_extensions.py:
```diff
@@ -0,0 +1,636 @@
+        self.assertIs(type(typing_extensions.Deque[int]()), collections.deque)
+        class D(typing_extensions.Deque[T]): ...
+        if TYPING_V3:
+            self.assertIs(type(D[int]()), D)
```
Potentially you might also add a test for ``__bases__`` they are also type-erased since one of the releases of ``typing``.

(But choose a class that actually derives from something subscripted, maybe ``Counter``.)

Michael0x2a replied:
I added a check for this using Counter, as you suggested.

ilevkivskyi commented on typing_extensions/src_py3/test_typing_extensions.py:
```diff
@@ -0,0 +1,636 @@
+
+
+if __name__ == '__main__':
+   # main(argv=[sys.argv[0]] + sys.argv[2:])
```
Is this comment necessary? Why do we need this pattern?

Michael0x2a replied:
This was a holdover from the code I previously had for running tests, which we removed. I removed the comment here + in the Python 2 tests.

ilevkivskyi commented on typing_extensions/src_py3/typing_extensions.py:
```diff
@@ -0,0 +1,612 @@
+try:
+    import collections.abc as collections_abc
+except ImportError:
+    import collections as collections_abc  # Fallback for PY3.2.
```
Again, this is not needed.

Michael0x2a replied:
Also removed.

ilevkivskyi commented on typing_extensions/src_py3/typing_extensions.py:
```diff
@@ -0,0 +1,612 @@
+    _generic_new = typing._generic_new
+else:
+    def _generic_new(base_cls, cls, *args, **kwargs):
+        return base_cls.__new__(cls, *args, **kwargs)
```
I am not 100% sure this is a safe substitution. Do you have a test for this? IIRC, the point was that ``__init__`` was not called at instantiation of subscripted generics.

Michael0x2a replied:
I don't know if I necessarily have tests for specifically this -- I was mostly copying what the typing module was already doing and assumed the behavior was correct. For example, in Python 3.5.0 - 3.5.2, the `List` type was defined like this:

```python
class List(list, MutableSequence[T], extra=list):
    def __new__(cls, *args, **kwds):
        if _geqv(cls, List):
            raise TypeError("Type List cannot be instantiated; "
                            "use list() instead")
        return list.__new__(cls, *args, **kwds)
```

In Python 3.5.3+, it was defined like this:

```python
class List(list, MutableSequence[T], extra=list):
    __slots__ = ()

    def __new__(cls, *args, **kwds):
        if _geqv(cls, List):
            raise TypeError("Type List cannot be instantiated; "
                            "use list() instead")
        return _generic_new(list, cls, *args, **kwds)
```

I'm also not sure if this substitution is 100% correct, but if it's broken, it'll at least be consistently broken in the same way the rest of the `typing` module was in Python 3.5.0 - 3.5.2. Defining the function this way also helped me reduce redundancy since I can define things like `Deque` just once instead of twice.

ilevkivskyi replied:
> be consistently broken in the same way the rest of the ``typing`` module was in Python 3.5.0 - 3.5.2

OK, but then please mention this in a docstring/comment.

ilevkivskyi commented on typing_extensions/src_py3/typing_extensions.py:
```diff
@@ -0,0 +1,612 @@
+    'NewType',
+    'overload',
+    'Text',
+    'TYPE_CHECKING',
```
Again, I am not sure about ``GenericMeta`` (see comment above).

ilevkivskyi commented:
Actually, I have one more suggestion. Maybe edit description to reflect the fact that experimental things that are going to appear in ``typing`` later will appear here first (like ``TypedDict``, argspecs, and ``Protocol``).

Michael0x2a commented:
@ilevkivskyi Thanks for the review! I'll start going through your comments and making fixes in a moment.

gvanrossum commented:
You will have to wait until after the release unless @JukkaL merges this
despite the code freeze.

On Jul 7, 2017 8:47 AM, "Michael Lee" <notifications@github.com> wrote:

> @ilevkivskyi <https://github.com/ilevkivskyi> Thanks for the review! I'll
> start going through your comments and making fixes in a moment.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/typing/pull/443#issuecomment-313719309>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACwrMvzaBjfKPMxJS4jdNXcCf6T9eISxks5sLlMkgaJpZM4OF__7>
> .
>

Michael0x2a commented on typing_extensions/setup.py:
```diff
@@ -0,0 +1,70 @@
+'''
+
+classifiers = [
+    'Development Status :: 5 - Production/Stable',
```
Done!

Michael0x2a commented on typing_extensions/setup.py:
```diff
@@ -0,0 +1,70 @@
+      author='Guido van Rossum, Jukka Lehtosalo, Lukasz Langa, Michael Lee',
+      author_email='jukka.lehtosalo@iki.fi',
+      # TODO: Change URL
+      url='https://github.com/michael0x2a/typing_extensions',
```
Whoops, forgot to change that.

Michael0x2a commented on tox.ini:
```diff
@@ -22,4 +22,9 @@ exclude =
+  src/test_typing.py,
+  typing_extensions/src_py2/test_typing_extensions.py,
+  typing_extensions/src_py3/test_typing_extensions.py,
+  # test_data should contain an exact snapshot of the standard
```
Same thing -- forgot to remove it.

Michael0x2a commented on .travis.yml:
```diff
@@ -14,5 +18,8 @@ install:
 
 script:
   - export PYTHONPATH=`python -c "import sys; print('python2' if sys.version.startswith('2') else 'src')"`; py.test $PYTHONPATH
+  - python setup.py install;
```
The problem is that the `typing_extensions` depends on `typing` being installed for Python 3.4 and below since we try and re-export as much as possible from the `typing` module. That said, I've tried adjusting travis so it should install `typing` only for Python 3.4 and below -- we'll see if I did it correctly once the tests are done.

Michael0x2a commented on typing_extensions/README.md:
```diff
@@ -0,0 +1,66 @@
+To run tests, navigate into the appropriate source directory and run
+`test_typing_extensions.py`. You will also need to install the latest
+version of `typing` if you are using a version of Python that does not
+include `typing` as a part of the standard library.
```
I added it to the `test_requirements.txt` file for the `typing` repo. If you prefer, I can leave that one alone and add a separate `test_requirements.txt` and `requirements.txt` file to the `typing_extensions` folder.

Since typing_extensions is basically a separate project, maybe it makes sense for it to have its own requirement files, but it does feel a little redundant...

ilevkivskyi replied:
I would actually prefer separate requirements for ``typing_extensions``.

Michael0x2a commented on typing_extensions/src_py2/test_typing_extensions.py:
```diff
@@ -0,0 +1,134 @@
+                message += ' : %s' % msg
+            raise self.failureException(message)
+
+    def clear_caches(self):
```
Removed.

Michael0x2a commented on typing_extensions/src_py2/test_typing_extensions.py:
```diff
@@ -0,0 +1,134 @@
+
+        cm = manager()
+        self.assertIsInstance(cm, typing_extensions.ContextManager)
+        self.assertNotIsInstance(42, typing_extensions.ContextManager)
```
Done!

Michael0x2a commented on typing_extensions/src_py2/test_typing_extensions.py:
```diff
@@ -0,0 +1,134 @@
+        self.assertIn('overload', a)
+        self.assertIn('Text', a)
+        self.assertIn('TYPE_CHECKING', a)
+
```
Removed.

Michael0x2a commented on typing_extensions/src_py2/test_typing_extensions.py:
```diff
@@ -0,0 +1,134 @@
+
+
+if __name__ == '__main__':
+    main(argv=[sys.argv[0]] + sys.argv[2:])
```
Done.

Michael0x2a commented on typing_extensions/setup.py:
```diff
@@ -0,0 +1,70 @@
+Typing -- Type Hints for Python
+
+This is a backport of the standard library typing module to Python
+versions 3.5+. The typing module has seen several changes since it was
```
Done! I just went with "Python 3.7" for the sake of simplicity.

JelleZijlstra commented on typing_extensions/src_py3/test_typing_extensions.py:
```diff
@@ -15,9 +15,29 @@
-TYPING_V3 = sys.version_info >= (3, 5, 3)
-TYPING_V4 = sys.version_info >= (3, 6, 1)
+TYPING_LATEST = sys.version_info[:3] < (3, 5, 0)
+TYPING_V2 = TYPING_LATEST or sys.version_info[:3] >= (3, 5, 1)
```
I think this would be clearer as `TYPING_3_5_1` or similar, because "v2" is not a terminology used elsewhere.

Michael0x2a replied:
Ok, done.

Michael0x2a commented:
@ilevkivskyi -- Ok, I should be ready for a second round of review. I tried adding links to relevant pull requests and issues for the typing version-specific if statements, but found it quite hard to do so in general, so unfortunately most of them are still undocumented :(.

Michael0x2a commented:
Just as an FYI, I decided to remove `typing_extensions.Collection` -- in retrospect, I think attempting to backport that type was a mistake.

In order to correctly backport that type, I basically need to inject `typing_extensions.Collection` into the middle of an existing inheritance hierarchy, which I don't think is possible to do cleanly, if at all. More concretely, the following code doesn't typecheck in Python 2 and older versions of Python 3 (the typing stubs have `typing.Sequence` subclass `typing.Collection` only some of the time):

    from mypy_extensions import Collection

    def foo(x):
        # type: Collection[int] -> int
        return sum(x)

    foo([1, 2, 3])

I think this is potentially repairable at least at typechecking-time with some very careful surgery on the typing stubs, but doing that seems questionable to me. It also wouldn't fix any runtime `isinstance` or `issubclass` checks people might try doing.

(More specifically, Python 3's typing stubs has `typing.Sequence` always subclass a phantom `_Collection` type, which is only publicly exposed as `typing.Collection` for Python 3.6+. We could add a similar phantom type to the Python 2 typing stubs, then have the `typing_extension` re-export `typing._Collection` as `typing_extensions.Collection`. Users would then need to use an up-to-date version of typeshed.)

Michael0x2a commented:
@ilevkivskyi -- as an update, I incorporated the changes you just pushed to master earlier today (removing `_gorg` and `_geqv` and making other speed improvements).

ilevkivskyi commented:
@Michael0x2a thanks for continuing to work on this! I will try to make the second review today or tomorrow.

ilevkivskyi commented:
@Michael0x2a 
I responded to your previous comments, continuing with review now.

ilevkivskyi commented:
Concerning ``Collection`` , I would still add it. Maybe not in this PR, but after you make an initial PR to typeshed with ``typing_extensions`` stubs.

ilevkivskyi commented on typing_extensions/README.rst:
```diff
@@ -0,0 +1,85 @@
+
+The ``typing_extensions`` module contains both backports of changes made to
+the ``typing`` module since Python 3.5.0 as well as experimental types that
+will be eventually added to the ``typing`` module once stabilized.
```
I think you forgot to mention that it is provisional until Python 3.7.

ilevkivskyi commented on typing_extensions/src_py2/test_typing_extensions.py:
```diff
@@ -0,0 +1,237 @@
+        self.assertIsSubclass(collections.Counter, Counter)
+        self.assertIs(type(Counter()), collections.Counter)
+        self.assertIs(type(Counter[T]()), collections.Counter)
+        self.assertIs(type(Counter[int]()), collections.Counter)
```
I actually wanted to also test that ``Counter.__bases__`` are already erased (before instantiation).

ilevkivskyi commented on typing_extensions/src_py3/test_typing_extensions.py:
```diff
@@ -0,0 +1,624 @@
+import typing_extensions
+import collections.abc as collections_abc
+
+TYPING_LATEST = sys.version_info[:3] < (3, 5, 0)
```
I am not sure I follow the logic here, shouldn't this be ``TYPING_OLDEST``? Or this is the version of typing you "emulate"? Could you please add a comment here?

Michael0x2a commented:
@ilevkivskyi -- ok, I've addressed your new comments and am ready for the 3rd round of review!

Regarding your comment regarding `Counter.__bases__`: I think I added a test for that in the Python 3 tests in response to one of your comments from the first round of review, but forgot to add the same test for the Python 2 tests. I copied the same check over.

That said, I want to make sure I actually understand both comments correctly -- are the tests I added what you were looking for, or should I need to add in more checks there?

Michael0x2a commented:
@ilevkivskyi stubs for typeshed can be found here: https://github.com/python/typeshed/pull/1471

I'll work on making changes to mypy itself next. iirc it contains some special-casing and tests specifically for mypy_extensions so I'll probably start by duplicating a subset of those for typing_extensions.
